services/horizon/internal/ingest/processors/contract_data.go

```diff
@@ -139,6 +150,106 @@ func AssetFromContractData(ledgerEntry xdr.LedgerEntry, passphrase string) *xdr.
+// https://github.com/stellar/rs-soroban-env/blob/5695440da452837555d8f7f259cc33341fdf07b0/soroban-env-host/src/native_contract/token/storage_types.rs#L11-L24
+func ContractBalanceFromContractData(ledgerEntry xdr.LedgerEntry, passphrase string) ([32]byte, *big.Int, bool) {
+	if ledgerEntry.Data.Type != xdr.LedgerEntryTypeContractData {
+		return [32]byte{}, nil, false
+	}
+
+	contractData := ledgerEntry.Data.MustContractData()
+	// we don't support asset stats for lumens
+	nativeAssetContractID, err := xdr.MustNewNativeAsset().ContractID(passphrase)
+	if err != nil || contractData.ContractId == nativeAssetContractID {
+		return [32]byte{}, nil, false
+	}
+
+	if contractData.Key.Type != xdr.ScValTypeScvObject {
+		return [32]byte{}, nil, false
+	}
+	keyObj := contractData.Key.MustObj()
+	if keyObj == nil {
+		return [32]byte{}, nil, false
+	}
+	if keyObj.Type != xdr.ScObjectTypeScoVec {
+		return [32]byte{}, nil, false
+	}
+	keyEnumVec := keyObj.MustVec()
+	if len(keyEnumVec) != 2 || !keyEnumVec[0].Equals(xdr.ScVal{Type: xdr.ScValTypeScvSymbol, Sym: &balanceMetadataSym}) {
+		return [32]byte{}, nil, false
+	}
+	if keyEnumVec[1].Type != xdr.ScValTypeScvObject {
+		return [32]byte{}, nil, false
+	}
+	addressObj := keyEnumVec[1].MustObj()
+	if addressObj == nil {
+		return [32]byte{}, nil, false
+	}
+	if addressObj.Type != xdr.ScObjectTypeScoAddress {
+		return [32]byte{}, nil, false
+	}
+	scAddress := addressObj.MustAddress()
+	if scAddress.Type != xdr.ScAddressTypeScAddressTypeContract {
+		return [32]byte{}, nil, false
+	}
+	holder := scAddress.MustContractId()
+
+	if contractData.Val.Type != xdr.ScValTypeScvObject {
+		return [32]byte{}, nil, false
+	}
+	obj := contractData.Val.MustObj()
+	if obj == nil {
+		return [32]byte{}, nil, false
+	}
+	if obj.Type != xdr.ScObjectTypeScoMap {
+		return [32]byte{}, nil, false
+	}
+	balanceMap := obj.MustMap()
+	if len(balanceMap) != 3 {
+		return [32]byte{}, nil, false
+	}
+
+	if balanceMap[0].Key.Type != xdr.ScValTypeScvSymbol ||
+		balanceMap[0].Key.MustSym() != "amount" ||
+		balanceMap[0].Val.Type != xdr.ScValTypeScvObject {
+		return [32]byte{}, nil, false
+	}
+	if balanceMap[1].Key.Type != xdr.ScValTypeScvSymbol ||
+		balanceMap[1].Key.MustSym() != "authorized" ||
+		!isBool(balanceMap[1].Val) {
+		return [32]byte{}, nil, false
+	}
+	if balanceMap[2].Key.Type != xdr.ScValTypeScvSymbol ||
+		balanceMap[2].Key.MustSym() != "clawback" ||
+		!isBool(balanceMap[1].Val) {
+		return [32]byte{}, nil, false
+	}
+	amountObj := balanceMap[0].Val.MustObj()
+	if amountObj == nil {
+		return [32]byte{}, nil, false
+	}
+	if amountObj.Type != xdr.ScObjectTypeScoI128 {
+		return [32]byte{}, nil, false
+	}
+	amount := amountObj.MustI128()
+	// check if negative
+	if ((amount.Hi >> 56) & 0x80) > 0 {
+		return [32]byte{}, nil, false
+	}
+	amt := new(big.Int).Lsh(big.NewInt(int64(amount.Hi)), 56)
+	amt = new(big.Int).Add(amt, big.NewInt(int64(amount.Lo)))
```

Comment:
same nit here as before, I think you can safely reuse `amt`:

```suggestion
	amt = amt.Add(amt, big.NewInt(int64(amount.Lo)))
```